File: Structural/Composite/CSharp/Folder.cs
// Folder.cs

using System;
using System.Collections.Generic;

public class Folder : IFileSystemComponent {
  private string name;
  private List<IFileSystemComponent> components = new List<IFileSystemComponent>();

  public Folder(string name) {
    this.name = name;
  }

  public void AddComponent(IFileSystemComponent component) {
    components.Add(component);
  }

  public void ShowDetails() {
    Console.WriteLine($"Folder: {name}");
    foreach (var component in components) {
      component.ShowDetails();
    }
  }
}

File: Structural/Composite/Cpp/Folder.cpp
// Folder.cpp

#include "Folder.h"

Folder::Folder(const std::string& name) : name(name) {}

void Folder::addComponent(std::unique_ptr<IFileSystemComponent> component) {
  components.push_back(std::move(component));
}

void Folder::showDetails() const {
  std::cout << "Folder: " << name << std::endl;
  for (const auto& component : components) {
    component->showDetails();
  }
}
